corehq/apps/sso/utils.py

```diff
@@ -23,3 +26,36 @@ def get_request_data(request):
         'get_data': request.GET.copy(),
         'post_data': request.POST.copy(),
     }
+
+
+def _get_full_sso_url(view_name, identity_provider):
+    return "{}{}".format(
+        get_url_base(),
+        reverse(view_name, args=(identity_provider.slug,))
+    )
+
+
+def get_saml_entity_id(identity_provider):
+    return _get_full_sso_url("sso_saml_metadata", identity_provider)
+
+
+def get_saml_acs_url(identity_provider):
+    return _get_full_sso_url("sso_saml_acs", identity_provider)
+
+
+def get_saml_sls_url(identity_provider):
+    return _get_full_sso_url("sso_saml_sls", identity_provider)
+
+
+def get_dashboard_link(identity_provider):
+    from corehq.apps.accounting.models import Subscription
+    from corehq.apps.sso.views.enterprise_admin import EditIdentityProviderEnterpriseView
+    linked_subscription = Subscription.visible_objects.filter(
+        account=identity_provider.owner,
+        is_active=True,
+        account__is_active=True,
+    ).first()
```

Comment:
use `get()`, not `first()`. `first()` is not semantically correct here. The reason I see some people prefer `first()` to `get()` is that they don't have to handle exceptions when the query returns nothing -- but that doesn't apply here, since the code is referencing into a possibly empty `linked_subscription`

Reply:
well, another reason is that a customer account might have multiple active subscriptions, so `get()` wouldn't necessarily work here and would throw an error on those type of accounts. :)

Reply:
Doh, you're 100% right. I wasn't reading that carefully enough, my bad.